tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c

```diff
@@ -0,0 +1,287 @@
+
+  switch (g_debug_sel) {
+
+    case DEBUG_SET_DCSR_STOPCOUNT:
+      __asm__ volatile ("csrrs x0, dcsr, %[set_dcsr_stopcount]" :: [set_dcsr_stopcount]"r"(set_dcsr_stopcount));
+      g_is_dcsr_stopcount = 1;
+      g_debug_sel = DEBUG_UNEXPECTED_ENTRY;
+      break;
+
+    case DEBUG_CLEAR_DCSR_STOPCOUNT:
+      __asm__ volatile ("csrrc x0, dcsr, %[set_dcsr_stopcount]" :: [set_dcsr_stopcount]"r"(set_dcsr_stopcount));
+      g_is_dcsr_stopcount = 0;
+      g_debug_sel = DEBUG_UNEXPECTED_ENTRY;
+      break;
+
+    case DEBUG_RUN_COUNTER_CHECK:
+      g_is_failure += counter3_minstret_tests(g_is_dcsr_stopcount);
+      g_debug_sel = DEBUG_UNEXPECTED_ENTRY;
+      break;
+
+    case DEBUG_UNEXPECTED_ENTRY:
+    default:
+      g_unexpected_debug_entry += 1;
+      break;
+  }
+  return;
+}
+
+/*
+ * execute_debug_command
+ *
+ * Sends commands debug handler
+ *
+ * Needed to execute commands that require to run with debug privelege
+ *
+ */
+void execute_debug_command(uint32_t dbg_cmd) {
+  // Disable trigger after use
+  g_debug_sel = dbg_cmd;
+
+  g_debug_entry_status = DEBUG_STATUS_NOT_ENTERED;
+  // Assert debug req
+  DEBUG_REQ_CONTROL_REG = (CV_VP_DEBUG_CONTROL_DBG_REQ(0x1)        |
+                           CV_VP_DEBUG_CONTROL_REQ_MODE(0x1)       |
+                           CV_VP_DEBUG_CONTROL_PULSE_DURATION(0x8) |
+                           CV_VP_DEBUG_CONTROL_START_DELAY(0xc8));
+  // Wait for debug entry
+  while (g_debug_entry_status == DEBUG_STATUS_NOT_ENTERED);
+}
+
+/*
+ * clear_counter3
+ *
+ * Clear mhpmcounter nr.  and verify that the counter is cleared
+ *
+ */
+int clear_counter3 (void) {
+
+  int counter = 0;
+
+  //Deactivate the counter
+  __asm__ volatile(R"(
+    mv t0, x0
+    not t0, t0
+    csrw mcountinhibit, t0
+  )"::: "t0");
+
+  //Clear the counter
+  __asm__ volatile(R"(
+    mv t0, x0
+    csrw 0xB03, t0
+  )"::: "t0");
+
+  //Read the counter's value
+  __asm__ volatile("csrr %[counter], 0xB03" : [counter]"=r"(counter));
+
+  if (counter != 0){
+    return 1;
+  } else {
+    return 0;
+  }
+
+}
+
+/*
+ * counter3_minstret_tests
+ *
+ * Configure the counter to count number of retired instructions,
+ *
+ * run a sequence with an expected number of instructions,
+ *
+ * and verify that mhpmcounter nr.  has counted, or not counted, the instructions according to DCSR's stopcount value.
+ *
+ */
+int counter3_minstret_tests (int g_is_dcsr_stopcount)
+  {
+  int is_failure = 0;
+  int counter = 0;
+  int cnt_retired_instr = (1 << MHPMEVENT_INSTR_BIT_POS);
+  int expected_minstret = 3 + 1; //3 nop instructions and 1 instruction to deactivate counter
+
+  //Make the counter count number of retired instructions
+  __asm__ volatile("csrw 0x323, %0 " :: "r"(cnt_retired_instr));
+
+  //Activate the counter, run some instructions, deactivate the counter
+  __asm__ volatile(R"(
+    mv t0, x0
+    not t1, t0
+    csrw mcountinhibit, t0
+    nop
+    nop
+    nop
+    csrw mcountinhibit, t1
+  )"::: "t0", "t1");
```

Comment:
I'm not 100% solid on the use of clobbers, but this does indeed match how I believe them to work.